execution_engine/src/core/runtime/mod.rs

```diff
@@ -1852,36 +1846,118 @@ where
+            None => return Err(Error::KeyNotFound(contract_key)),
         };
 
         let entry_point = contract
             .entry_point(&entry_point_name)
             .cloned()
             .ok_or_else(|| Error::NoSuchMethod(entry_point_name.to_owned()))?;
 
+        self.call_contract_checked(contract_package, contract_hash, contract, entry_point, args)
+    }
+
+    /// Calls contract if caller has access, and args match entry point definition
+    fn call_contract_checked(
+        &mut self,
+        contract_package: ContractPackage,
+        contract_hash: ContractHash,
+        contract: Contract,
+        entry_point: EntryPoint,
+        args: RuntimeArgs,
+    ) -> Result<CLValue, Error> {
+        // if public, allowed
+        // if not public, restricted to user group access
         self.validate_entry_point_access(&contract_package, entry_point.access())?;
 
-        for (expected, found) in entry_point
-            .args()
-            .iter()
-            .map(|a| a.cl_type())
-            .cloned()
-            .zip(args.to_values().into_iter().map(|v| v.cl_type()).cloned())
+        // This will skip arguments check for system contracts only. This code should be removed on
+        // next major version bump. Argument checks for system contract is still done during
+        // execution of a system contract.
+        if !self
+            .protocol_data()
+            .system_contracts()
+            .contains(&contract_hash)
         {
-            if expected != found {
-                return Err(Error::type_mismatch(expected, found));
+            let required_args: BTreeSet<&str> = entry_point
+                .args()
+                .iter()
+                .filter_map(|param| {
+                    if !param.cl_type().is_option() {
+                        Some(param.name())
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            let optional_args: BTreeSet<&str> = entry_point
+                .args()
+                .iter()
+                .filter_map(|param| {
+                    if param.cl_type().is_option() {
+                        Some(param.name())
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            let passed_args: BTreeSet<&str> = args
+                .named_args()
+                .map(|named_arg| named_arg.name())
+                .collect();
+
+            // Unused args are the ones that are passed, but are not present as required nor
+            // optionals.
+            let unused_args = &(&passed_args - &required_args) - &optional_args;
+
+            if !unused_args.is_empty() {
+                return Err(Error::UnusedArgumentsFound {
+                    required: required_args.len(),
+                    unused: unused_args.len(),
+                });
+            }
```

Comment:
I don't think it is safe for us to error in this case; some 3rd parties may tail their passed args with extra marker or readback data that is not needed for the called logic but is used for some off chain post process triggered by inclusion of the deploy in a block.

Reply:
Fair point. I removed this error condition, and added an extra test to ensure contracts can accept arbitrary arguments (not listed in entry point as extra markers; for example)